### PR TITLE
feat: reset inputs after submission

### PR DESCRIPTION
**Pull request by zzmp**
Zeroes out the inputs and sets Field.INPUT as the independent field (ie TradeType.EXACT_INPUT) once a swap or wrap is submitted. This is in line with github.com/Uniswap/interface, as per @infredible.

https://user-images.githubusercontent.com/5403956/186292639-13744922-1b1e-4fb4-b5ee-d5b477fd88c5.mov

### CODE REVIEW

File: src/components/Swap/SwapButton/index.tsx [hunk 1]
```diff
@@ -67,80 +67,73 @@ export default memo(function SwapButton({ disabled }: SwapButtonProps) {
...
-  const addTransaction = useAddTransaction()
+  const addTransactionInfo = useAddTransaction()
   const setDisplayTxHash = useUpdateAtom(displayTxHashAtom)
-  const setOldestValidBlock = useSetOldestValidBlock()
+  const [, setSwapAmount] = useSwapAmount(Field.INPUT)
+  const resetSwapAmount = useCallback(() => setSwapAmount(''), [setSwapAmount])
+
+  const onSubmit = useCallback(
+    async (submit: () => Promise<TransactionInfo | undefined>): Promise<void> => {
+      try {
```
Comment (vm) [inline src/components/Swap/SwapButton/index.tsx:77]: which part of this actually throws an error? would be better to wrap only that piece

File: src/components/Swap/SwapButton/index.tsx [hunk 2]
```diff
@@ -67,80 +67,73 @@ export default memo(function SwapButton({ disabled }: SwapButtonProps) {
...
-      if (isAnimating(document)) {
-        document.addEventListener('animationend', postWrap)
-      } else {
-        setIsPending(false)
-      }
-    } catch (e) {
-      // TODO(zzmp): Surface errors from wrap.
-      console.log(e)
-    } finally {
-      setIsPending(false)
```
Comment (vm) [inline src/components/Swap/SwapButton/index.tsx:103]: is this not necessary anymore?

Comment (zzmp) [reply src/components/Swap/SwapButton/index.tsx:103]: `onSubmit` will always return, so a `finally` is no longer needed.

File: src/components/Swap/SwapButton/index.tsx [hunk 3]
```diff
@@ -67,80 +67,73 @@ export default memo(function SwapButton({ disabled }: SwapButtonProps) {
...
-      }
-    } catch (e) {
-      // TODO(zzmp): Surface errors from swap.
-      console.log(e)
-    }
-  }, [addTransaction, setDisplayTxHash, setOldestValidBlock, swapCallback, trade.trade])
+      invariant(trade.trade)
+      return { type: TransactionType.SWAP, response, tradeType: trade.trade.tradeType, trade: trade.trade }
+    })
+    setOpen(false)
```
Comment (zzmp) [inline src/components/Swap/SwapButton/index.tsx]: This should only be called in the case of a successful submission.

File: src/components/Swap/SwapButton/index.tsx (hunk 1 above)
Comment (zzmp) [reply src/components/Swap/SwapButton/index.tsx:79]: The submission. Done.

File: src/components/Swap/SwapButton/index.tsx [hunk 4]
```diff
@@ -67,80 +67,82 @@ export default memo(function SwapButton({ disabled }: SwapButtonProps) {
...
-        response: transaction,
-        tradeType: trade.trade.tradeType,
-        trade: trade.trade,
-      })
-      setDisplayTxHash(transaction.hash)
+    const submitted = await onSubmit(async () => {
+      const response = await swapCallback?.()
+      if (!response) return
 
       // Set the block containing the response to the oldest valid block to ensure that the
```
Comment (JFrankfurt) [inline src/components/Swap/SwapButton/index.tsx:131]: shouldn't all these `oldest` references be `newest`?

File: src/components/Swap/SwapButton/index.tsx [hunk 5]
```diff
@@ -67,80 +67,82 @@ export default memo(function SwapButton({ disabled }: SwapButtonProps) {
...
-      }
-      if (isAnimating(document)) {
-        document.addEventListener('animationend', postSwap)
-      } else {
-        setOpen(false)
-      }
-    } catch (e) {
-      // TODO(zzmp): Surface errors from swap.
-      console.log(e)
+      invariant(trade.trade)
```
Comment (JFrankfurt) [inline src/components/Swap/SwapButton/index.tsx:137]: is adding an `invariant` call here the correct move? iirc it throws if the argument is falsy, so this would be an uncaught error and trigger the error boundary.

File: src/components/Swap/SwapButton/index.tsx [hunk 6]
```diff
@@ -67,80 +67,82 @@ export default memo(function SwapButton({ disabled }: SwapButtonProps) {
   // Close the review modal on chain change.
   useEffect(() => setOpen(false), [chainId])
 
-  const addTransaction = useAddTransaction()
+  const addTransactionInfo = useAddTransaction()
   const setDisplayTxHash = useUpdateAtom(displayTxHashAtom)
-  const setOldestValidBlock = useSetOldestValidBlock()
+  const [, setSwapAmount] = useSwapAmount(Field.INPUT)
+  const resetSwapAmount = useCallback(() => setSwapAmount(''), [setSwapAmount])
```
Comment (JFrankfurt) [inline src/components/Swap/SwapButton/index.tsx:73]: I was looking for *where* the the `Field.INPUT` setting was happening, then realized it's probably happening implicitly here. Just want to point out that this feels a little hard to parse. I don't know if someone reading this code w/o context would be able to recognize that is happening.

File: src/components/Swap/SwapButton/index.tsx [hunk 7]
```diff
@@ -67,80 +67,82 @@ export default memo(function SwapButton({ disabled }: SwapButtonProps) {
   // Close the review modal on chain change.
   useEffect(() => setOpen(false), [chainId])
 
-  const addTransaction = useAddTransaction()
+  const addTransactionInfo = useAddTransaction()
```
Comment (JFrankfurt) [inline src/components/Swap/SwapButton/index.tsx:70]: should this hook name change?

File: src/components/Swap/SwapButton/index.tsx [hunk 8]
```diff
@@ -67,80 +67,82 @@ export default memo(function SwapButton({ disabled }: SwapButtonProps) {
...
-      console.log(e)
-    } finally {
-      setIsPending(false)
-    }
-  }, [addTransaction, native, setDisplayTxHash, wrapCallback, wrapType])
+    await onSubmit(async () => {
+      const response = await wrapCallback()
+      if (!response) return
+
+      invariant(wrapType)
```
Comment (JFrankfurt) [inline src/components/Swap/SwapButton/index.tsx:114]: This is a subtly different assertion than the one being removed. `WrapType.NONE` (`0`) will throw here now. Is that intentional?

**Review by JFrankfurt — APPROVED** [review]

File: src/components/Swap/SwapButton/index.tsx (hunk 7 above)
Comment (zzmp) [reply src/components/Swap/SwapButton/index.tsx:70]: This is adding a `TransactionInfo` type (`Transaction` type is different), so I'm going to propagate this change to all callsites.

File: src/components/Swap/SwapButton/index.tsx (hunk 8 above)
Comment (zzmp) [reply src/components/Swap/SwapButton/index.tsx:114]: `WrapType.NONE` was removed in https://github.com/Uniswap/widgets/pull/159 - this is now a type alias for `TransactionType.WRAP | TransaactionType.UNWRAP | undefined`.

I can (and should) still make this more explicit, so I'll update to `wrapType !== undefined`.

File: src/components/Swap/SwapButton/index.tsx (hunk 6 above)
Comment (zzmp) [reply src/components/Swap/SwapButton/index.tsx:73]: I thought that by semantically naming a derived setter I would make it easier to parse, so this is great feedback. I'll opt for the more explicit `setSwapAmount('')` inlined, and alias `setSwapAmount` to `setInputAmount` or similar.

File: src/components/Swap/SwapButton/index.tsx (hunk 4 above)
Comment (zzmp) [reply src/components/Swap/SwapButton/index.tsx:131]: we want the `oldest` valid block because that implies that nothing older than that is still valid.
I'm happy to change this to something less ambiguous in a separate PR, but would prefer not to change the naming in this PR.

File: src/components/Swap/SwapButton/index.tsx (hunk 5 above)
Comment (zzmp) [reply src/components/Swap/SwapButton/index.tsx:137]: `onWrap` will not be called if `wrapType === undefined`, but due to the complex type that's not evident to the type checker. We _could_ replace this with a check, but if `wrapCallback()` returns a valid response, then `wrapType` should also be valid. I can annotate this with a comment.

File: src/components/Swap/SwapButton/index.tsx (hunk 6 above)
Comment (JFrankfurt) [reply src/components/Swap/SwapButton/index.tsx]: specifically I mean it's not immediately clear to me that these lines reset the widget to an `exactInput` swap if it was not immediately in that state. The derived setter makes sense for the `Amount` for sure, just not the `Field`/`tradeType`
